File: src/main/java/dev/lonami/uniffidl/UdlDocumentationProvider.java
package dev.lonami.uniffidl;

import com.intellij.lang.documentation.AbstractDocumentationProvider;
import com.intellij.lang.documentation.DocumentationMarkup;
import com.intellij.psi.PsiElement;
import com.intellij.psi.presentation.java.SymbolPresentationUtil;
import dev.lonami.uniffidl.psi.UdlDefinition;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class UdlDocumentationProvider extends AbstractDocumentationProvider {
    @Override
    public @Nullable @Nls String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
        if (element instanceof UdlDefinition) {
            UdlDefinition definition = (UdlDefinition) element;
            final String name = definition.getName();
            final String file = SymbolPresentationUtil.getFilePathPresentation(element.getContainingFile());
            final String docComment = UdlUtil.findDocumentationComment(definition);
            return renderFullDoc(UdlUtil.getDefinitionTypeText(definition), name, file, docComment);
        }
        return null;
    }

    @Override
    public @Nullable @Nls String generateHoverDoc(@NotNull PsiElement element, @Nullable PsiElement originalElement) {
        return generateDoc(element, originalElement);
    }

    @Override
    public @Nullable @Nls String getQuickNavigateInfo(PsiElement element, PsiElement originalElement) {
        if (element instanceof UdlDefinition) {
            UdlDefinition definition = (UdlDefinition) element;
            final String name = definition.getName();
            final String file = SymbolPresentationUtil.getFilePathPresentation(element.getContainingFile());
            return "'" + name + "' in " + file;
        }
        return null;
    }

    private String renderFullDoc(String type, String name, String file, String docComment) {
        StringBuilder sb = new StringBuilder();
        sb.append(DocumentationMarkup.DEFINITION_START);
        sb.append(type);
        sb.append(DocumentationMarkup.DEFINITION_END);
        sb.append(DocumentationMarkup.CONTENT_START);
        sb.append(name);
        sb.append(DocumentationMarkup.CONTENT_END);
        sb.append(DocumentationMarkup.SECTIONS_START);
        addKeyValueSection("Name:", name, sb);
        addKeyValueSection("File:", file, sb);
        addKeyValueSection("Comment:", docComment, sb);
        sb.append(DocumentationMarkup.SECTIONS_END);
        return sb.toString();
    }

    private void addKeyValueSection(String key, String value, StringBuilder sb) {
        sb.append(DocumentationMarkup.SECTION_HEADER_START);
        sb.append(key);
        sb.append(DocumentationMarkup.SECTION_SEPARATOR);
        sb.append("<p>");
        sb.append(value);
        sb.append(DocumentationMarkup.SECTION_END);
    }
}

File: src/main/kotlin/dev/lonami/uniffidl/UdlDocumentationProvider.kt
package dev.lonami.uniffidl

import com.intellij.lang.documentation.AbstractDocumentationProvider
import com.intellij.lang.documentation.DocumentationMarkup
import com.intellij.psi.PsiElement
import com.intellij.psi.presentation.java.SymbolPresentationUtil
import dev.lonami.uniffidl.psi.UdlDefinition
import org.jetbrains.annotations.Nls

class UdlDocumentationProvider : AbstractDocumentationProvider() {
    override fun generateDoc(element: PsiElement?, originalElement: PsiElement?): @Nls String? {
        if (element is UdlDefinition) {
            val name = element.name
            val file = SymbolPresentationUtil.getFilePathPresentation(element.containingFile)
            val docComment = UdlUtil.findDocumentationComment(element)
            return renderFullDoc(UdlUtil.getDefinitionTypeText(element), name, file, docComment)
        }
        return null
    }

    override fun generateHoverDoc(element: PsiElement, originalElement: PsiElement?): @Nls String? {
        return generateDoc(element, originalElement)
    }

    override fun getQuickNavigateInfo(element: PsiElement?, originalElement: PsiElement?): @Nls String? {
        if (element is UdlDefinition) {
            val name = element.name
            val file = SymbolPresentationUtil.getFilePathPresentation(element.containingFile)
            return "'$name' in $file"
        }
        return null
    }

    private fun renderFullDoc(type: String, name: String, file: String, docComment: String): String {
        val sb = StringBuilder()
        sb.append(DocumentationMarkup.DEFINITION_START)
        sb.append(type)
        sb.append(DocumentationMarkup.DEFINITION_END)
        sb.append(DocumentationMarkup.CONTENT_START)
        sb.append(name)
        sb.append(DocumentationMarkup.CONTENT_END)
        sb.append(DocumentationMarkup.SECTIONS_START)
        addKeyValueSection("Name:", name, sb)
        addKeyValueSection("File:", file, sb)
        addKeyValueSection("Comment:", docComment, sb)
        sb.append(DocumentationMarkup.SECTIONS_END)
        return sb.toString()
    }

    private fun addKeyValueSection(key: String, value: String, sb: StringBuilder) {
        sb.append(DocumentationMarkup.SECTION_HEADER_START)
        sb.append(key)
        sb.append(DocumentationMarkup.SECTION_SEPARATOR)
        sb.append("<p>")
        sb.append(value)
        sb.append(DocumentationMarkup.SECTION_END)
    }
}
